### PR TITLE
Reject HTTP requests early

### PR DESCRIPTION
**Pull request by LevitatingBusinessMan**
As a safety measure to prevent CSRF attacks.
See:
https://github.com/mopidy/mopidy-mpd/issues/18
https://github.com/mopidy/mopidy/issues/1659

### CODE REVIEW

**Comment by kingosticks** [comment]
Since this is specifically an MPD protocol thing I'd be tempted to move all of this to `MpdSession.on_line_received()` i.e. after it has been decoded and split into lines. This allows the request to be logged and for the MPD error message (assuming there is one rather than just closing the connection) to be sent to the client.

**Comment by kingosticks** [comment]
It also needs a test and for the existing tests that broke due to this, to be fixed.

**Review by kingosticks — CHANGES_REQUESTED** [review]

File: mopidy_mpd/network.py [hunk 1]
```diff
@@ -454,6 +454,14 @@ def on_stop(self):
         self.connection.stop("Actor is shutting down.")
 
     def parse_lines(self):
+        """
+        All mpd commands begin with a lowercase letter
+        so we can immediately reject any potential HTTP request.
+        """
+        if self.recv_buffer[0] < ord('a') or self.recv_buffer[0] > ord('z'):
+            self.connection.stop("Invalid mpd command")
```
[the remark this answers is not in the document]
Comment (LevitatingBusinessMan) [reply mopidy_mpd/network.py]: https://github.com/MusicPlayerDaemon/MPD/blob/72621531e0a540fd4d0363c25f8be216a3a0154f/src/client/Process.cxx#L60-L68

I currently made it respond "ERR malformed command", but I have to dig a bit into the framework to figure out how to replicate the error like in mpd.

File: mopidy_mpd/network.py [hunk 2]
```diff
@@ -454,6 +454,14 @@ def on_stop(self):
         self.connection.stop("Actor is shutting down.")
 
     def parse_lines(self):
+        """
```
[the remark this answers is not in the document]
Comment (LevitatingBusinessMan) [reply mopidy_mpd/network.py]: My bad, it's been a while since I've worked with python.

File: mopidy_mpd/session.py [hunk 3]
```diff
@@ -29,6 +29,14 @@ def on_start(self):
     def on_line_received(self, line):
         logger.debug("Request from %s: %s", self.connection, line)
 
+        # To prevent CSRF attacks, lines with an invalid command syntax are immediately dropped.
+        command = line.split(" ")[0]
```
Comment (kingosticks) [inline mopidy_mpd/session.py:33]: Why do we need to split?

Comment (LevitatingBusinessMan) [reply mopidy_mpd/session.py:33]: I only check against the actual command name, as the possible characters for that are known. So I split the line at the spaces and then grab the first element (the command name) to check the validity.

But I think the protocol actually supports any whitespace between commands and arguments so I should probably change this.

Comment (kingosticks) [reply mopidy_mpd/session.py:33]: The first character in the line is also the first character of the command.

**Comment by LevitatingBusinessMan** [comment]
@kingosticks I just checked, and `mpd` just drops connections starting with an invalid character outright. No errors are returned.

So I'll leave it up to you to decide what error we should return if any.

File: mopidy_mpd/network.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kingosticks) [reply mopidy_mpd/network.py]: I'm not familiar with what mpd's "client" is. Normally we run mpd and see what it does.

File: mopidy_mpd/session.py (hunk 3 above)
Comment (LevitatingBusinessMan) [reply mopidy_mpd/session.py]: @kingosticks I chose validate the complete command name here. But I could return to just testing the first character like `mpd` if that's preferred.

Comment (kingosticks) [reply mopidy_mpd/session.py]: Yes, I think that's better rather than trying to do more complicated parsing here.

File: mopidy_mpd/session.py [hunk 4]
```diff
@@ -29,6 +29,15 @@ def on_start(self):
     def on_line_received(self, line):
         logger.debug("Request from %s: %s", self.connection, line)
 
+        # To prevent CSRF attacks, requests with an invalid command syntax are immediately dropped.
+        if len(line.split()) > 0:
+            command = line.split()[0]
+            if not all(c.islower() and c.isalpha() or c == '_' for c in list(command)):
+                logger.debug("Dropping connection due to malformed command")
+                self.send_lines([f"ERR malformed command {command}"])
```
Comment (kingosticks) [inline mopidy_mpd/session.py:37]: >mpd just drops connections starting with an invalid character outright. No errors are returned.

Perfect, lets get rid of this and so the same.

**Comment by LevitatingBusinessMan** [comment]
@kingosticks `test_empty_request` is currently erroring and I will fix that. But I want to note that the behavior tested with `test_empty_request` isn't the same behavior anymore seen in `mpd`, since the test was written before `mpd` started dropping malformed requests. So maybe `test_empty_request` should do `self.assertNoResponse()` instead of expecting an error.

**Comment by kingosticks** [comment]
My version (Music Player Daemon 0.21.5 (0.21.5)) behaves the same as `test_empty_request`.

**Comment by LevitatingBusinessMan** [comment]
> My version (Music Player Daemon 0.21.5 (0.21.5)) behaves the same as `test_empty_request`.

On `0.23.5` is if use netcat to open a connection and I send just whitespace the connection immediately closes. Don't forget that `0.21.5` is almost 3 years old.

**Comment by LevitatingBusinessMan** [comment]
I modified `test_empty_request` to reflect what `mpd` seems to be doing. If you object to that I will just make sure the previous expectations are satisfied thought it might make the added check a bit more complicated. Otherwise I think this PR should be ready?

**Comment by kingosticks** [comment]
Yeh I do appreciate debian stable is *very* stable! Our compatibility is all over the place right now so it's fine either way. I did take a quick look through mpd's changelog but nothing jumped out at me.

**Review by kingosticks — APPROVED** [review]

**Review by jodal — APPROVED** [review]
Thanks for fixing this!

**Comment by kingosticks** [comment]
Merged in https://github.com/mopidy/mopidy-mpd/commit/c6b0d4c9f985c7e79dc910ca517f4efabb9054d1

**Pull request closed without merging**